subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java

```diff
@@ -349,7 +378,15 @@ public boolean shouldCleanupOutputsOnNonIncrementalExecution() {
 
         @Override
         public Optional<? extends Iterable<String>> getChangingOutputs() {
-            return Optional.empty();
+            if (!vfsEnabled) {
+                return Optional.empty();
+            }
+            ImmutableList.Builder<String> builder = ImmutableList.builder();
+            visitOutputProperties((propertyName, type, roots) -> roots.forEach(
+                root -> builder.add(root.getAbsolutePath())
+            ));
+            context.getTaskProperties().getDestroyableFiles().forEach(file -> builder.add(file.getAbsolutePath()));
+            return Optional.of(builder.build());
```

Comment:
What about local state?

Reply:
Good question. I suppose we need to invalidate this as well.